src/main/groovy/org/ajoberstar/grgit/auth/JschAgentProxySessionFactory.java

```diff
@@ -0,0 +1,186 @@
+import com.jcraft.jsch.agentproxy.usocket.NCUSocketFactory;
+
+/**
+ * A session factory that supports use of ssh-agent and Pageant SSH authentication.
+ *
+ * @since 0.1.0
+ */
+public class JschAgentProxySessionFactory extends JschConfigSessionFactory {
+  private static final Logger logger = LoggerFactory.getLogger(JschAgentProxySessionFactory.class);
+  private final AuthConfig config;
+
+  public JschAgentProxySessionFactory(AuthConfig config) {
+    this.config = config;
+  }
+
+  /** Customize session */
+  @Override
+  protected void configure(Host hc, Session session) {
+    config.getSessionConfig().forEach(session::setConfig);
+  }
+
+  /**
+   * Obtains a JSch used for creating sessions, with the addition of ssh-agent and Pageant agents,
+   * if available.
+   *
+   * @return the JSch instance
+   */
+  @Override
+  protected JSch getJSch(Host hc, FS fs) throws JSchException {
+    JSch jsch;
+    try {
+      jsch = super.getJSch(hc, fs);
+    } catch (JSchException e) {
+      jsch = super.createDefaultJSch(fs);
+    }
+
+    if (config.getSshPrivateKeyPath() != null) {
+      if (config.getSshPassphrase() != null) {
+        jsch.addIdentity(config.getSshPrivateKeyPath(), config.getSshPassphrase());
+      } else {
+        jsch.addIdentity(config.getSshPrivateKeyPath());
+      }
+    }
+
+    Optional<Connector> con = determineConnector();
+    if (con.isPresent()) {
+      IdentityRepository remoteRepo = new RemoteIdentityRepository(con.get());
+      if (remoteRepo.getIdentities().isEmpty()) {
+        logger.info("not using agent proxy: no identities found");
+      } else {
+        logger.info("using agent proxy");
+        jsch.setIdentityRepository(remoteRepo);
+      }
+    } else {
+      logger.info("jsch agent proxy not available");
+    }
+    return jsch;
+  }
+
+  /**
+   * Chooses which agent proxy connector is used.
+   *
+   * @return the connector available at this time
+   */
+  private Optional<Connector> determineConnector() {
+    return Stream.<Supplier<Optional<Connector>>>of(this::sshAgentSelector, this::pageantSelector)
+        .map(Supplier::get)
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .findFirst();
+  }
+
+  private Optional<Connector> sshAgentSelector() {
+    try {
+      if (!config.allows(AuthConfig.Option.SSHAGENT)) {
+        logger.info("ssh-agent option disabled");
+        return Optional.empty();
+      } else if (SSHAgentConnector.isConnectorAvailable()) {
+        Optional<USocketFactory> usf = determineUSocketFactory();
+        if (usf.isPresent()) {
+          logger.info("ssh-agent available");
+          return Optional.of(new SSHAgentConnector(usf.get()));
+        } else {
+          logger.info("ssh-agent not available");
+          return Optional.empty();
+        }
+      } else {
+        logger.info("ssh-agent not available");
+        return Optional.empty();
+      }
+    } catch (Throwable e) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Catch Exception instead of Throwable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.ajoberstar.com/coding_rules#rule_key=squid%3AS1181)